# **IRIS DATASET PROEJCT**


## Exploring the Data

First of all, we will import the libraries we will be using into our code. The ones we will be using here are the numpy and pandas. 

In [ ]:
import numpy as np
import pandas as pd

To proceed forward in this project, first we will import the scikit-learn iris dataset, a classic example of machine learning and statastics. By using load_iris function we will load the dataset.

In [ ]:
from sklearn import datasets
iris = datasets.load_iris()

Like a dictionary, the iris dataset has keys and values.

In [ ]:
print("Keys of Iris datasets: \n{}".format(iris.keys()))

Keys of Iris datasets: 
dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names', 'filename'])


Let's discuss these keys in details and understand what type of data do they store.
1. **DESCR**

    The DESCR stores a short summary about the iris dataset, hence by using the print string we can find further about the dataset.




In [ ]:
print(iris['DESCR'][:200] + "\n . . .") #with the 200 value we are able to view upto 200 words of the data.

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, predictive
 . . .


  2. **target_names**
      
      It is an array of strings of species of flowers that we want to predicti.e. 'setosa', 'versicolor' and 'virginica'.

In [ ]:
print("Target names: {}".format(iris['target_names']))

Target names: ['setosa' 'versicolor' 'virginica']


  3. **feature_names**

        This is a list of strings which includes each feature of description of the flowers i.e. 'sepal length (cm)', 'sepal width (cm)', 'petal length (cm)' and 'petal width (cm)'

In [ ]:
print("Feature names: \n{}".format(iris['feature_names']))

Feature names: 
['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


  4. **data**

        This contains the numeric values (the training data) of sepal length,sepal width, petal length and petal width in NumPy array. 

In [ ]:
print("Type of data: {}". format(type(iris['data'])))

Type of data: <class 'numpy.ndarray'>


  Rows in the data correspond to the flowers while the columns correspond to the four measurements(feature names) of the flowers.

In [ ]:
print("shape of the data: {}".format(iris['data'].shape))

shape of the data: (150, 4)


  From the above code it can be observed that the array holds the feature measurement for 150 different flowers. Here the 150 flowers can be categorised as the data points. The sahpe of the data s 150 sample multiplied by 4 features.

In [ ]:
print("First five rows of the data: {}".format(iris['data'][:5]))

First five rows of the data: [[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]]


  The first colum represents the the sepal length. Second reprsents the sepal width and thrid and fourth represent petal lengths and petals widths respectively.

  5. **target**

        Target is a one dimensional array that contains one entry per flower.        

In [ ]:
print("Type of target: {}".format(type(iris['target'])))

Type of target: <class 'numpy.ndarray'>


In [ ]:
print("Shape of target: {}".format(iris['target'].shape))

Shape of target: (150,)


  Based on the target values we can conclude that the flowers are encoded as integers from 0 to 2.

In [ ]:
print("Target: \n{}".format(iris['target']))

Target: 
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]


  Here the '0' represents the 'setosa', '1' represents 'versicolor' and the '2' represents 'virginica'.

This represent the filename of the project.

In [ ]:
print("File name {}".format(iris['filename']))

File name /usr/local/lib/python3.6/dist-packages/sklearn/datasets/data/iris.csv


## Measuring Sucess: Training and Testing Data

In scikit-learn, data is usually denoted with a capital X, while labels are denoted by lower case y. THis is inspired by standard formulation in mathematics, f(x)=y, where x is input and y is the output by convention.
Capital X is for data, a two-dimensional array, and lower case y is for target, a one-dimensional array (a vector).

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
                        iris['data'], iris['target'], random_state=0)

Classically in a machine learning model, the 75% should be training data whereas 25% should be test data.

In [ ]:
print("X_train shape: {}".format(X_train.shape))
print("y_train shape: {}".format(y_train.shape))
print("X_test shape: {}".format(X_test.shape))
print("y_test shape: {}".format(y_test. shape))

X_train shape: (112, 4)
y_train shape: (112,)
X_test shape: (38, 4)
y_test shape: (38,)


Here we have printed the shape of the X-train, X-test, y_train, y_test. Both the training data and test data have same column respectively.

Now we are making a k- nearest neighbor learning model.

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=1)

Now after giving the data we need to train the model.

In [ ]:
knn.fit(X_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=1, p=2,
                     weights='uniform')

## Making Predictions

Now that the model has been trained, we can now start making predictions on new data and provide it an appropriate label.

So, now let's test the data and put it in a NumPy array to make a prediction about the label. Assume that sepal length is 5 cm, 2.9 cm is sepal width, 1 cm is petal length and 0.2 cm is petal width.

In [ ]:
X_new = np.array([[5, 2.9, 1, 0.2]])
print("X_new.shape: {}".format(X_new.shape))
X_new.shape: (1, 4)

X_new.shape: (1, 4)


Now, using the model we are going to predict the species of the flower (target name).

In [ ]:
prediction = knn.predict(X_new)

print("Prediction: {}".format(prediction))
print("Predict target name: {}".format(
    iris['target_names'][prediction]))


Prediction: [0]
Predict target name: ['setosa']


## Evaluating the Model

Now we are going to test the model and check its efficiency.

In [ ]:
y_pred = knn.predict(X_test)
print("Test set predictions: \n {}".format(y_pred))

Test set predictions: 
 [2 1 0 2 0 2 0 1 1 1 2 1 1 1 1 0 1 1 0 0 2 1 0 0 2 0 0 1 1 0 2 1 0 2 2 1 0
 2]


In [ ]:
print("Test set score: {}".format(np.mean(y_pred == y_test)))

Test set score: 0.9736842105263158


Hence, accuracy of our model is 97.368%.